programmer-skills/big-O-notation.md

```diff
@@ -0,0 +1,161 @@
+(N-1) + (N-2) + (N-3)...+ 2+1 = N(N-1)/2.
+
+```javascript
+function isUnique(array)
+{
+	for (var i = 0; i < array.length-1; i++)
+	{
+		for (var j = i+1; j< array.length; j++)
+		{
+			if (array[i]==array[j] ) return false;
+		}
+	}
+	return true;
+}
+```
+
+###O(2<sup>N</sup>)
+O(2<sup>N</sup>) corresponds to an algorithm where adding a  new item to the data set doubles the number of steps.
+In case of O(2<sup>N</sup>) algrithms the number of steps grow exponentially.
+
+A well-known example is the naive recursive calculation of the Fibonnaci-series.
+In general:
+f(n) = f(n-1) + f(n-2) = f(n-2) + f(n-3) + f(n-3) + f(n-4) =
+= f(n-3) + f(n-4) + f(n-4) + f(n-5) + f(n-4) + f(n-5) + f(n-5) + f(n-6) = ...
+
+```javascript
+	function fibonacci(n)
+	{
+		if (n===0 || n===1) return 1;
+		else return ( fibonacci(n-1) + fibonacci(n-2) );
+	}
+```
+
+###O(log N)
+O(log N) algorithms are highly efficient for handling huge data sets. It has little effect on the runtime whether or data set has a 1000 or 10 000 entries.
+
+A well known example of O(log N) algorithms is binary search,  it is used to find items
+in already sorted data sets. Binary search is a good algorithm for finding 
+names in the phone book, or for the egg dropping problem(There is a building of 100 floors. If an egg is dropped from the Nth floor or above it will break. 
+If it’s dropped from any floor below, it won't break. Find N!). 
+
+Binary search looks for a target value by comparing the middle item of an array.
+
+Binary search algorithm:   
+1. find the middle element,  
+2. if the middle item equals to the target value, then stop,  
+3. Otherwise:  
+	* the target value is smaller than the middle element. Repeat from step 1 for the left-subarray, before the middle element,  
+	* the target value is bigger than the middle element. Repeat from step 1 for the right-subarray, after the middle element.  
+
+	
+Using binary search we halve the data set in every iteration circle. Thus if one doubles the size of the data set, it has very little effect on the runtime. 
+Considering worst case scenarios, to find an element
+ in a twice as big data set as the orignal requires just one more iteration cycle.
+	
+```javascript
+function binarySearch(array, target)
+{
+	var start = 0;
+	var stop = array.length -1;
+	var middleIndex;
+		while (start <= stop)
+		{
+		middleIndex = Math.floor( (start+stop)/2 );
+		console.log(middleIndex);
+		if (array[middleIndex] == target){
+			return middleIndex;
+		}else if (array[middleIndex] > target )
+		{
+			stop = middleIndex-1;
+		}else
+		{
+			start = middleIndex+1;
+		}
+	}
+	return -1;
+}
+```
+
+###Summary comparison of different runtimes:
+![bigO](http://i.stack.imgur.com/ia6VB.png) 
+ 
+*Fig. 1 is taken from [Stackoverflow](http://stackoverflow.com/questions/4317414/polynomial-time-and-exponential-time)*
+
+###References
```

Comment:
Might be nice to use Markdown link notation to give names to the links and hide the URLs
e.g. [Big-O notations for common data structures and algorithms](http://bigocheatsheet.com/)